File: works/tmp/15558.py
from collections import deque


def bfs(n, k):
    mark = [[False for _ in range(200010)] for _ in range(2)]
    q = deque()
    q.append([0,0,0])

    while q:
        left = q.popleft()
        r = left[0]
        c = left[1]
        time = left[2]

        if c >= n: return True
        mark[r][c] = True

        if c+1 >= time+1 and c+1 >= n: return True
        if c+1 >= time+1 and arr[r][c+1] != '0' and mark[r][c+1] == False:
            mark[r][c+1] = True
            q.append([r, c+1, time+1])

        if c-1 >= time+1 and c-1 >= 0 and arr[r][c-1] != '0' and mark[r][c-1] == False:
            mark[r][c-1] = True
            q.append([r,c-1,time+1])

        nr = 1 if r == 0 else 0
        if c+k >= time+1 and c+k >= n: return True
        if c+k >= time+1 and arr[nr][c+k] != '0' and mark[nr][c+k] == False:
            mark[nr][c+k] = True
            q.append([nr, c+k, time+1])

    return False


n, k = map(int, input().split())
arr = []
s = input()
arr.append(s)
s = input()
arr.append(s)

ans = bfs(n, k)
if ans: print(1)
else: print(0)

File: works/tmp/15558.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.IO;
using static System.Console;

namespace ConsoleApp
{
    class Tuple
    {
        public int r, c, time;
        public Tuple(int _r, int _c, int _time) { r = _r; c = _c; time = _time; }
    }

    class Program
    {
        static int n, k;
        static bool[,] mark = new bool[2, 200010];
        static string[] arr = new string[2];

        static bool bfs()
        {
            Queue q = new Queue();
            q.Enqueue(new Tuple(0, 0, 0));

            while (q.Count != 0)
            {
                Tuple top = (Tuple)q.Dequeue();
                int r = top.r;
                int c = top.c;
                int time = top.time;

                if (c >= n) return true;
                mark[r, c] = true;

                if (c + 1 >= time + 1 && c + 1 >= n) return true;
                if (c + 1 >= time + 1 && arr[r][c + 1] != '0' && !mark[r, c + 1])
                {
                    mark[r, c + 1] = true;
                    q.Enqueue(new Tuple(r, c + 1, time + 1));
                }

                if (c - 1 >= time + 1 && c - 1 >= 0 && arr[r][c - 1] != '0' && !mark[r, c - 1])
                {
                    mark[r, c - 1] = true;
                    q.Enqueue(new Tuple(r, c - 1, time + 1));
                }

                int nr = r == 0 ? 1 : 0;
                if (c + k >= time + 1 && c + k >= n) return true;
                if (c + k >= time + 1 && arr[nr][c + k] != '0' && !mark[nr, c + k])
                {
                    mark[nr, c + k] = true;
                    q.Enqueue(new Tuple(nr, c + k, time + 1));
                }
            }

            return false;
        }

        static void Main(string[] args)
        {
            StreamReader sr = new StreamReader(Console.OpenStandardInput());
            string[] _s = sr.ReadLine().Split(' ');
            n = int.Parse(_s[0]);
            k = int.Parse(_s[1]);

            for (int i = 0; i < 2; i++)
            {
                string s = sr.ReadLine(); arr[i] = s;
            }

            int ans = 0;
            if (bfs()) ans = 1;
            Write(ans);
        }

    }
}
